src/OpenTelemetry.AutoInstrumentation/Instrumentations/NoCode/NoCodeIntegrationHelper.cs

```diff
@@ -0,0 +1,140 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+using System.Diagnostics;
+using System.Reflection;
+using OpenTelemetry.AutoInstrumentation.CallTarget;
+using OpenTelemetry.AutoInstrumentation.Configurations;
+using OpenTelemetry.AutoInstrumentation.Util;
+
+namespace OpenTelemetry.AutoInstrumentation.Instrumentations.NoCode;
+
+internal static class NoCodeIntegrationHelper
+{
+    private static readonly ActivitySource Source = new("OpenTelemetry.AutoInstrumentation.NoCode");
+
+    internal static List<NoCodeInstrumentedMethod> NoCodeEntries { get; set; } = [];
+
+    internal static CallTargetState OnMethodBegin()
+    {
+        const int methodNameFrameIndex = 3;
+
+        var method = new StackFrame(methodNameFrameIndex).GetMethod();
+        var methodName = method?.Name;
+        var typeName = method?.DeclaringType?.FullName;
+        var assemblyName = method?.DeclaringType?.Assembly.GetName().Name;
+        var parameters = method?.GetParameters()!;
+        var noCodeEntry = NoCodeEntries.Single(x =>
+            x.Definition.TargetType == typeName &&
+            x.Definition.TargetMethod == methodName &&
+            x.Definition.TargetAssembly == assemblyName &&
+            CheckParameters(x.SignatureTypes, parameters));
+
+        // TODO Span kind and attributes from configuration
+        // TODO Consider execute some dynamic code to build name/span/attributes based on config and taking data from method parameters
+        var activity = Source.StartActivity(name: noCodeEntry.SpanName, kind: ActivityKind.Internal, tags: [new("TODO", "attributes-from-configuration")]);
+        return new CallTargetState(activity, null);
+    }
+
+    internal static CallTargetReturn<TReturn> OnMethodEnd<TReturn>(TReturn returnValue, Exception? exception, in CallTargetState state)
+    {
+        var activity = state.Activity;
+        if (activity is null)
+        {
+            return new CallTargetReturn<TReturn>(returnValue);
+        }
+
+        var returnType = typeof(TReturn);
+        if (returnType.IsGenericType)
+        {
+            var genericReturnType = returnType.GetGenericTypeDefinition();
+            if (typeof(Task).IsAssignableFrom(returnType))
+            {
+                // The type is a Task<>
+                return new CallTargetReturn<TReturn>(returnValue);
+            }
+#if NET
+
+            if (genericReturnType == typeof(ValueTask<>))
```

Comment:
ValueTasks can be used in net framework apps if the appropriate libraries are added to the application. Why is this check in a NET compiler directive?

Reply:
In the same way we have implemented it in the standard bytecode instrumentation. I just mimic this behavior. https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/283fbb7149a3ece491c2099e66a35a307718cfe3/src/OpenTelemetry.AutoInstrumentation/CallTarget/Handlers/EndMethodHandler%601.cs#L34-L63

If needed, we should update it in both places.

Do you think that we need this at this moment/schedule some issue for follow up/vNext?

Reply:
#4502